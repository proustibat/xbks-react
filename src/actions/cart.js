export const addBook = book => ( {
    type: 'ADD_BOOK',
    book
} );

export const removeBook = ( isbn = '' ) => ( {
    type: 'REMOVE_BOOK',
    isbn
} );

// Could be used if cart is saved in localstorage or cookie
export const setSavedBooks = books => ( {
    type: 'SET_SAVE_BOOKS',
    books
} );

export const changeQuantity = ( { isbn, qtt } ) => ( {
    type: 'CHANGE_QUANTITY',
    isbn,
    qtt
} );
