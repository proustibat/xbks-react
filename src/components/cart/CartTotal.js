import React from 'react';
import { connect } from 'react-redux';
import getSubTotal from '../../selectors/getSubTotal';
import getBestDiscount from '../../selectors/getBestDiscount';
import numeral from "numeral";

const CartTotal = props => (
    <div className="cart-totals">
        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Total before reductions:</span>
            <span className="cart-totals__line-amount">{ numeral( props.subtotal ).format( '0.00$' ) }</span>
        </p>
        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Reductions <em>(type: { props.discount.type })</em>:</span>
            <span className="cart-totals__line-amount">{ numeral( props.discount.amount ).format( '0.00$' ) }</span>
        </p>

        <p className="cart-totals__line">
            <span className="cart-totals__line-label">Total to pay:</span>
            <span className="cart-totals__line-amount">{ numeral( props.total ).format( '0.00$' ) }</span>
        </p>
    </div>
);

const mapStateToProps = state => ( {
    subtotal: getSubTotal( state.cart.books ),
    discount: state.offers.bestOffer,
    total: getSubTotal( state.cart.books ) - state.offers.bestOffer.amount
} );

export default connect( mapStateToProps )( CartTotal );
