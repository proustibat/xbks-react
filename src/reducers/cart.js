const cartReducerDefaultState = {
    books: []
};

export default ( state = cartReducerDefaultState, action ) => {
    let books;
    switch ( action.type ) {
        case 'ADD_BOOK_TO_CART':
            const existingBook = state.books.find( book => book.isbn === action.book.isbn );
            if ( existingBook ) {
                existingBook.quantity = existingBook.quantity + 1;
                books = state.books;
            }
            else {
                action.book.quantity = 1;
                books = [ ...state.books, action.book ];
            }
            return { ...state, ...{ books } };
        case 'REMOVE_BOOK_FROM_CART':
            books = state.books.filter( ( { isbn } ) => action.isbn !== isbn );
            return { ...state, ...{ books } };
        // case 'SET_SAVED_BOOKS_TO_CART':
        //     action.books.forEach( book => {
        //         if ( !book.quantity ) {
        //             book.quantity = 1;
        //         }
        //     } );
        //     return { ...state, ...{ books: action.books } };
        case 'CHANGE_QUANTITY_BOOK_IN_CART':
            const booksList = [...state.books];
            let newBookList = booksList
                .map( book => {
                    book.quantity = book.isbn === action.isbn ? action.qtt : book.quantity;
                    return book;
                } )
                .filter( book => book.quantity > 0 );
            return { ...state, ...{ books: newBookList } };
        default:
            return state;
    }
};
