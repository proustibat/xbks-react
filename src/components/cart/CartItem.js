import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import { changeQuantity, removeBook } from '../../actions/cart';
import { startSetOffers } from "../../actions/offers";

export class CartItem extends React.Component {

    state = {
      subTotal: 0
    };

    constructor( props ) {
        super( props );
        this.state = {
            isbn: props.isbn || '',
            title: props.title || '',
            price: props.price || 0,
            quantity: props.quantity || 0
        }
    }

    handleNumberChange = ( e ) => {
        console.log( this.props.isbn, e.currentTarget.value );
        const quantity = parseInt( e.currentTarget.value, 10 );

        this.setState( () => ( { quantity } ) );

        this.props.changeQuantity( {
            isbn: this.state.isbn,
            qtt: quantity
        } );
        this.props.startSetOffers();
    };

    handleRemove = () => {
      this.props.removeBook( this.state.isbn );
      this.props.startSetOffers();
    };

    render() {
        return (
            <div className="cart-list-item">
            <span className="cart-list__col-item">
                <Link className="cart-list__col-item-title" to={ `/book/${ this.state.isbn }` }>{ this.state.title }</Link>
            </span>
            <span className="cart-list__col-qtt">
                <button className="cart-list__delete" onClick={ this.handleRemove }>Remove</button>
                <input type="number" min="0" id="itemsNumber" defaultValue={ this.state.quantity } step="1" onInput={ this.handleNumberChange } />
            </span>
                <span className="cart-list__col-unit-price">{ numeral( this.state.price ).format( '0.00$') }</span>
                <span className="cart-list__col-sub-total">{ numeral( this.state.quantity * this.state.price ).format( '0.00$' ) }</span>
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => ( {
    changeQuantity: ( { isbn, qtt } ) => dispatch( changeQuantity( { isbn, qtt } ) ),
    removeBook: isbn => dispatch( removeBook( isbn ) ),
    startSetOffers: () => dispatch ( startSetOffers() )
} );

export default connect( undefined, mapDispatchToProps )( CartItem );
