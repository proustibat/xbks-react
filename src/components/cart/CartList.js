import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';

const CartList = props => (
    <div className="cart-list">
        { props.books.length === 0 ? (
            <span>Your cart is empty</span>
        ) : (
            <div>
                <div className="cart-list-header">
                    <span className="cart-list__col-item">Item</span>
                    <span className="cart-list__col-qtt">Qtt</span>
                    <span className="cart-list__col-unit-price">Unit Price</span>
                    <span className="cart-list__col-sub-total">Sub Total</span>
                </div>
                <div className="cart-list-body">
                    { props.books.map( ( book, i ) => (
                        <CartItem key = { `${ book.isbn }-${ i }` } { ...book } />
                    ) ) }
                </div>
            </div>
        ) }
    </div>
);

const mapStateToProps = state => ( {
    books: state.cart.books
} );

export default connect( mapStateToProps )( CartList);
