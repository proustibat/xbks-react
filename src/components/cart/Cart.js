import React from 'react';
import CartList from './CartList';
import CartTotal from './CartTotal';

export class Cart extends React.Component {
    render () {
        return (
            <div className="cart">
                <CartList />
                <CartTotal />
            </div>
        );
    }
}

export default Cart;
