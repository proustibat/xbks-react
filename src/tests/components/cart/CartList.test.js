import React from 'react';
import { shallow } from 'enzyme';
import { CartList } from "../../../components/cart/CartList";
import cart from "../../fixtures/cart";

test( 'Should render CartList component correctly with books', () => {
    const wrapper = shallow( <CartList
        books = { cart.books }
    /> );
    expect( wrapper ).toMatchSnapshot();
} );

test( 'Should render CartList component correctly with no books', () => {
    const wrapper = shallow( <CartList
        books = { [] }
    /> );
    expect( wrapper ).toMatchSnapshot();
} );
