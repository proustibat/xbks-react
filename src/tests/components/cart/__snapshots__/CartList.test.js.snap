// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CartList component correctly with books 1`] = `
<div
  className="cart-list"
>
  <div>
    <div
      className="cart-list-header"
    >
      <span
        className="cart-list__col-item"
      >
        Item
      </span>
      <span
        className="cart-list__col-qtt"
      >
        Qtt
      </span>
      <span
        className="cart-list__col-unit-price"
      >
        Unit Price
      </span>
      <span
        className="cart-list__col-sub-total"
      >
        Sub Total
      </span>
    </div>
    <div
      className="cart-list-body"
    >
      <Connect(CartItem)
        cover="http://henri-potier.xebia.fr/hp0.jpg"
        isbn="c8fabf68-8374-48fe-a7ea-a00ccd07afff"
        key="c8fabf68-8374-48fe-a7ea-a00ccd07afff-0"
        price={35}
        quantity={1}
        synopsis={
          Array [
            "Après la mort de ses parents (Lily et James Potier), Henri est recueilli par sa tante Pétunia (la sœur de Lily) et son oncle Vernon à l'âge d'un an. Ces derniers, animés depuis toujours d'une haine féroce envers les parents du garçon qu'ils qualifient de gens « bizarres », voire de « monstres », traitent froidement leur neveu et demeurent indifférents aux humiliations que leur fils Dudley lui fait subir. Henri ignore tout de l'histoire de ses parents, si ce n'est qu'ils ont été tués dans un accident de voiture",
            "Le jour des 11 ans de Henri, un demi-géant du nom de Rubeus Hagrid vient le chercher pour l’emmener à Poudlard, une école de sorcellerie, où il est inscrit depuis sa naissance et attendu pour la prochaine rentrée. Hagrid lui révèle alors qu’il a toujours été un sorcier, tout comme l'étaient ses parents, tués en réalité par le plus puissant mage noir du monde de la sorcellerie, Voldemort (surnommé « Celui-Dont-On-Ne-Doit-Pas-Prononcer-Le-Nom »), après qu'ils ont refusé de se joindre à lui. Ce serait Henri lui-même, alors qu'il n'était encore qu'un bébé, qui aurait fait ricocher le sortilège que Voldemort lui destinait, neutralisant ses pouvoirs et le réduisant à l'état de créature quasi-insignifiante. Le fait d'avoir vécu son enfance chez son oncle et sa tante dépourvus de pouvoirs magiques lui a donc permis de grandir à l'abri de la notoriété qu'il a dans le monde des sorciers.",
            "Henri entre donc à l’école de Poudlard, dirigée par le professeur Albus Dumbledore. Il est envoyé dans la maison Gryffondor par le « choixpeau ». Il y fait la connaissance de Ron Weasley et Hermione Granger, qui deviendront ses complices. Par ailleurs, Henri intègre rapidement l'équipe de Quidditch de sa maison, un sport collectif très populaire chez les sorciers se pratiquant sur des balais volants. Henri connaît probablement la plus heureuse année de sa vie, mais également la plus périlleuse, car Voldemort n'a pas totalement disparu et semble bien décidé à reprendre forme humaine.",
          ]
        }
        title="Henri Potier à l'école des sorciers"
      />
      <Connect(CartItem)
        cover="http://henri-potier.xebia.fr/hp1.jpg"
        isbn="a460afed-e5e7-4e39-a39d-c885c05db861"
        key="a460afed-e5e7-4e39-a39d-c885c05db861-1"
        price={30}
        quantity={1}
        synopsis={
          Array [
            "Henri Potier passe l'été chez les Dursley et reçoit la visite de Dobby, un elfe de maison. Celui-ci vient l'avertir que des évènements étranges vont bientôt se produire à Poudlard et lui conseille donc vivement de ne pas y retourner. Henri choisit d'ignorer cet avertissement. Le jour de son départ pour l'école, il se retrouve bloqué avec Ron Weasley à la gare de King's Cross, sans pouvoir se rendre sur le quai 9 ¾ où les attend le Poudlard Express. En dernier recours, les garçons se rendent donc à Poudlard à l'aide de la voiture volante de Monsieur Weasley et manquent de peu de se faire renvoyer dès leur arrivée à l'école pour avoir été aperçus au cours de leur voyage par plusieurs moldus.",
            "Le nouveau professeur de défense contre les forces du mal, Gilderoy Lockhart, se montre particulièrement narcissique et inefficace. Pendant ce temps, Henri entend une voix étrange en parcourant les couloirs du château, systématiquement associée à la pétrification immédiate d'un élève moldu de l'école. Dès la première attaque, un message sanglant apparaît sur l'un des murs, informant que la Chambre des secrets a été ouverte. Dumbledore et les autres professeurs (ainsi que Henri, Ron et Hermione) doivent prendre les mesures nécessaires pour trouver l'identité du coupable et protéger les élèves contre de nouvelles agressions.",
          ]
        }
        title="Henri Potier et la Chambre des secrets"
      />
    </div>
  </div>
</div>
`;

exports[`Should render CartList component correctly with no books 1`] = `
<div
  className="cart-list"
>
  <p
    className="cart-list__empty-message"
  >
    Your cart is empty
  </p>
</div>
`;
